# Домашнее задание 2 (5 баллов).

*Все задания ниже имеют равный вес (5/10)*

Код для импорта мы написали за вас (не благодарите, нам не трудно). Дальше код будете писать вы. 

[Тут](https://habr.com/ru/companies/ruvds/articles/494720/) шпора по pandas. За основу домашнего задания взят ноутбук [отсюда](https://rutube.ru/video/f884aa6ed5f94120b7304506042fe5bb/) (не подглядывайте!).

In [1]:
import pandas as pd
import numpy as np

#### Описание данных

Автор д/з - плохой человек, который не стал переводить описание с мотивировкой, что весь DS на английском. Так что описание полей будет на английском:

1. Account ID
- Description: A unique identifier for each social media account in the dataset.
- Type: Integer
- Example: 1, 2, 3, …
2. Username
- Description: The username or handle of the social media account.
- Type: String
- Example: john_doe, tech_guru_22, fitness_freak
3. Platform
- Description: The social media platform the account is using (Instagram, Twitter, Facebook, TikTok, LinkedIn).
- Type: Categorical (String)
- Example: Instagram, Twitter, Facebook, TikTok, LinkedIn
4. Follower Count
- Description: The total number of followers the account has.
- Type: Integer
- Example: 1500, 245000, 78000
5. Posts Per Week
- Description: The average number of posts the account creates per week.
- Type: Integer
- Example: 3, 5, 7
6. Engagement Rate
- Description: The percentage of interactions (likes, comments, shares) relative to the follower count. This is a measure of how engaging the content is.
- Type: Float
- Range: 0.01 to 0.15
- Example: 0.045 (4.5% engagement rate)
7. Ad Spend (USD)
- Description: The monthly amount spent on advertising or promoting posts.
- Type: Float
- Example: 150.75, 850.00, 300.50
8. Conversion Rate
- Description: The percentage of users who take a desired action (e.g., clicking a link, signing up, etc.) after interacting with an ad.
- Type: Float
- Range: 0.01 to 0.05 (1% to 5% conversion rate)
- Example: 0.025 (2.5% conversion rate)
9. Campaign Reach
- Description: The total number of unique users reached by the user’s campaigns in a given month.
- Type: Integer
- Example: 5000, 20000, 15000

#### Задание 0

Подгрузите данные. Да-да, за чтение таблицы баллов не будет))

**Hint**: [pd.read_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html)

In [2]:
df = pd.read_csv('data.csv') # ^⨀ᴥ⨀^

In [4]:
df = pd.read_csv("data.csv")
df

Account ID       Username   Platform  Follower Count  Posts Per Week  \
0              1     harrislisa     TikTok           54217               3   
1              2         rhicks   LinkedIn          987518               5   
2              3        qthomas   Facebook          218870               3   
3              4     carlosholt  Instagram          207432               6   
4              5  parsonsashley   Facebook          350204               2   
...          ...            ...        ...             ...             ...   
9995        9996          sbyrd     TikTok          388169               6   
9996        9997  jacksongerald     TikTok          150775               1   
9997        9998          eryan    Twitter          427451               5   
9998        9999      barbara57   LinkedIn          132884               4   
9999       10000        nbutler   Facebook          725253               2   

      Engagement Rate  Ad Spend (USD)  Conversion Rate  Campaign Reach  
0              0.0986          538.10           0.0490            1308  
1              0.0834          479.24           0.0174           13302  
2              0.1020          150.36           0.0318           11043  
3              0.0834          932.62           0.0400           12074  
4              0.0642          504.44           0.0463           14083  
...               ...             ...              ...             ...  
9995           0.0642          362.91           0.0108           45091  
9996           0.0642          177.92           0.0262           33084  
9997           0.0834          432.65           0.0402           12484  
9998           0.0986          892.99           0.0329           35766  
9999           0.0642          878.87           0.0420           13012  

[10000 rows x 9 columns]

#### Задание 1

Колонка `Platform` содержит название различных платформ. Давайте представим, что в них есть некоторое отношение порядка. Закодируйте каждую платформу целым числом (от 0 до N) и положите этот "код" в новую колонку `Platform_Code`. Теперь вычислите корреляцию Спирмена между всеми парами колонок в датасете (результатом будет таблица корреляций). В качестве ответа выведите значение корреляции `Platform_Code` с `Engagement Rate`. Можете после вывода числа еще коротко написать, что оно означает (нет, это не оценивается).

**Hint**: [pd.factorize](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.factorize.html), [pd.DataFrame.select_dtypes](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.select_dtypes.html), [pd.DataFrame.corr](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.corr.html).

In [5]:
from scipy.stats import spearmanr

In [6]:
# ( ੭ ･ᴗ･ )੭

df['Platform_Code'] = pd.factorize(df['Platform'])[0]

numeric_df = df.select_dtypes(include=['number'])

corr_matrix = numeric_df.corr(method='spearman')

print("Таблица корреляции Спирмена между всеми парами колонок в датасете,")
print('(таблица корреляций):')
print(corr_matrix)

platform_engagement_corr = corr_matrix.loc['Platform_Code', 'Engagement Rate']
print("\nКорреляция между Platform_Code и Engagement Rate:", platform_engagement_corr)


Таблица корреляции Спирмена между всеми парами колонок в датасете,
(таблица корреляций):
                 Account ID  Follower Count  Posts Per Week  Engagement Rate  \
Account ID         1.000000        0.002866       -0.003058         0.000611   
Follower Count     0.002866        1.000000       -0.004853        -0.016428   
Posts Per Week    -0.003058       -0.004853        1.000000         0.016908   
Engagement Rate    0.000611       -0.016428        0.016908         1.000000   
Ad Spend (USD)    -0.004928       -0.004516       -0.001298         0.007712   
Conversion Rate    0.005095        0.014002       -0.011176        -0.006964   
Campaign Reach    -0.006896       -0.020816        0.003364         0.013299   
Platform_Code     -0.005208        0.010304       -0.011840         0.031382   

                 Ad Spend (USD)  Conversion Rate  Campaign Reach  \
Account ID            -0.004928         0.005095       -0.006896   
Follower Count        -0.004516         0.014002      

#### Задание 2

Теперь посмотрите на столбец `Follower Count`. В нем какие-то числа. Иногда бывает полезно провести дискретизацию такого признака. Разбейте все значения в столбце на 4 группы: "Low", "Medium", "High", "Very High". Каждая группа включает в себя новые 25% данных. То есть, Low включает в себя 25% самых маленьких значений признака и так далее. Положите значения "Low", "Medium", "High" или "Very High" для каждого сэмпла датасета в новую колонку `Follower_Bin`. Теперь посчитайте среднее значение `Engagement Rate` для каждой категории из `Follower_Bin`. В качестве ответа выведите значение для категории "High".

**Hint**: [pd.qcut](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html), [pd.groupby](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html), [pd.DataFrame.mean](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.mean.html)

In [7]:
# (◕^^◕)

df['Follower_Bin'] = pd.qcut(df['Follower Count'], q=4, labels=["Low", "Medium", "High", "Very High"])

mean_engagement = df.groupby('Follower_Bin')['Engagement Rate'].mean()

high_engagement_rate = mean_engagement['High']

high_engagement_rate


C:\Users\chern\AppData\Local\Temp\ipykernel_4180\3181041049.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  mean_engagement = df.groupby('Follower_Bin')['Engagement Rate'].mean()


0.08655032

#### Задание 3

Иногда бывает полезно превратить широкую таблицу в длинную (например, для визуализаций сразу нескольких признаков на одной картинке). Да, звучит странно, но именно этим вы сейчас и займетесь. Сделайте новый датафрейм `melted_df`, в который вы поместите каждый сэмпл датасета 6 раз: по одному разу на значение из 'Follower Count', 'Posts Per Week', 'Ad Spend (USD)', 'Conversion Rate', 'Engagement Rate' и 'Campaign Reach'. То есть, вы берете сэмпл из датасета (строку) и превращаете ее в 6 отдельных строк. Каждая отдельная строка в столбце `Metric` имеет имя из предложенного списка 5 признаков, а в столбце `Value` - значение данного сэмпла по этому признаку. Значение `Platform` повторяется в этих 6 строках.

Иначе говоря, 

```json
{
    "Account ID": 1,
    "Username": "harrislisa",
    "Platform": "TikTok",
    "Follower Count": 54217,
    "Posts Per Week": 3,
    "Engagement Rate": 0.0986,
    "Ad Spend (USD)": 538.1,
    "Conversion Rate": 0.049,
    "Campaign Reach": 1308,
    "Platform_Code": 0,
    "Follower_Bin": "Low"
}
```

превращается в 

```json
{
    "Platform": "TikTok",
    "Metric": "Follower Count",
    "Value": 54217,
},
{
    "Platform": "TikTok",
    "Metric": "Posts Per Week",
    "Value": 3,
}, ...
```

Для каждого уникальной пары значений (`Platform`, `Metric`) посчитайте моду среди всех значений `Value` для этой пары, результат сделайте списком и оставьте только наибольшее. В качестве ответа выведите сумму полученных мод (сумму всех значений в столбце `Value` уже после вычисления мод). Иначе говоря, выведите сумму всех мод значений для всех уникальных пар (`Platform`, `Metric`).

**Hint**: [pd.melt](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.melt.html), [pd.DataFrame.mode](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.mode.html), [pd.DataFrameGroupBy.agg](https://pandas.pydata.org/docs/dev/reference/api/pandas.core.groupby.DataFrameGroupBy.agg.html)

In [8]:
# (づ๑•ᴗ•๑)づ♡
melted_df = pd.melt(df, id_vars=['Platform'], 
                        value_vars=['Follower Count', 'Posts Per Week', 'Ad Spend (USD)', 'Conversion Rate', 'Engagement Rate', 'Campaign Reach'],
                        var_name='Metric',
                        value_name='Value')
            
mode_df = melted_df.groupby(['Platform', 'Metric'])['Value'].agg(lambda x: x.mode().max()).reset_index()

total_mode_sum = mode_df['Value'].sum()

total_mode_sum

3100285.4716

#### Задание 4

А теперь хочется посмотреть на самые популярные аккаунты на разных платформах. Для каждой платформы отсортируйте датафрейм по убыванию количества подписчиков (`Follower Count`) - да, без циклов, сразу для всех платформ сделать сортировку, а затем оставьте только первые три записи для каждой платформы - это и будут три самых популярных аккаунта для каждой платформы. В качестве ответа выведите саму таблицу и минимальное значение `Follower Count` в ней.

**Hint**: к *groupby* можно применять функции - это эквивалентно применению функции к каждой "группе" внутри groupby-объекта. Читайте [про применение apply к датафрейму после groupby](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html#flexible-apply).

In [9]:
# ε(´סּ︵סּ`)з

top_accounts = df.sort_values(by='Follower Count', ascending=False).groupby('Platform').head(3)

print(top_accounts)

min_follower_count = top_accounts['Follower Count'].min()
print("Минимальное значение Follower Count:", min_follower_count)

      Account ID         Username   Platform  Follower Count  Posts Per Week  \
2403        2404           eric65   Facebook          999982               6   
4920        4921      teresaellis    Twitter          999919               6   
5838        5839     edwardthomas     TikTok          999739               7   
8685        8686  alexandersamuel  Instagram          999726               3   
9684        9685           sriley    Twitter          999442               3   
3965        3966         lrodgers  Instagram          999351               1   
4234        4235    andradewesley     TikTok          999234               5   
3039        3040          toneill   LinkedIn          999055               4   
6359        6360    andrewgregory   LinkedIn          998968               7   
2159        2160     ashleycooper   LinkedIn          998925               6   
2575        2576     williamwyatt     TikTok          998623               6   
7576        7577       peggymunoz    Twi

#### Задание 5

Хочется посчитать какую-то метрику. Мы хотим посмотреть, на отношение разности суммы подписчиков аккаунтов с высокой и низкой конверсией к суммарному охвату рекламы на каждой платформе. То есть, мы делим аккаунты на две группы: высокая и низка конверсия. Затем мы смотрим на то, на сколько сильно влияние аккаунтов с высокой конверсией по сравнению с аккаунтами с низкой конверсией. 

Давайте определим *Conversion Influence* следущим образом:

$$Conversion Influence = \frac{Total Follower\ Count (High) - Total Follower\ Count (Low)}{Total Campaign Reach (High)+Total Campaign Reach (Low)}$$

Считать эту метрику мы будет для каждой `Platform`. В этой формуле High - это значения всех сэмплов датасета, в которых `Conversion Rate` больше медианы, а `Low` - не более медианы. `Total Feature` - это суммарное количество значений `Feature` либо по `High` сэмплам, либо по `Low`.

Чтобы постоянно не пересчитывать, где High. где Low, сделайте новую колонку в датасете `Conversion_Category`. Положите в нее для каждой строки либо High, либо Low.

Выведите платформу с самым большим `Conversion Influence`.

**Hint**: данное задание не про *groupby*, а скорее про [pd.pivot_table](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html). Сделайте сводную таблицу, по которой уже можно посчитать суммы, а затем подставить их в формулы.

In [10]:
# (︶ω︶)

median_conversion_rate = df['Conversion Rate'].median()

df['Conversion_Category'] = df['Conversion Rate'].apply(lambda x: 'High' if x > median_conversion_rate else 'Low')

pivot_table = pd.pivot_table(df, 
                             values=['Follower Count', 'Campaign Reach'], 
                             index='Platform', 
                             columns='Conversion_Category', 
                             aggfunc='sum', 
                             fill_value=0)

pivot_table['Conversion Influence'] = (pivot_table['Follower Count']['High'] - pivot_table['Follower Count']['Low']) / \
                                      (pivot_table['Campaign Reach']['High'] + pivot_table['Campaign Reach']['Low'])

max_influence_platform = pivot_table['Conversion Influence'].idxmax()
max_influence_value = pivot_table['Conversion Influence'].max()

print(f"Платформа с самым большим Conversion Influence: {max_influence_platform} ({max_influence_value})")

Платформа с самым большим Conversion Influence: Twitter (1.0908724546727178)


#### Задание 6

Мы знаем, что вам понравилось считать метрики по формуле. Давайте закрепим этот успех. Теперь для каждой платформы посчитаем, на сколько эффективна реклама в разрезе трех последовательных записей в датасете. 

Для каждой платформы отсортируйте записи в порядке убывания `Posts Per Week`. Будто бы аккаунты, которые постят чаще, используют более "активные" стратегии по рекламе. Теперь посчитайте *скользущие суммы с окном 3* по `Campaign Reach` и `Ad Spend (USD)`. Скользящая сумма с окном N - это вы идете по массиву, берете все последовательные тройки записей и суммируете их. Для первых двух записей троек не найдется. Для них скользящее среднее - NaN, что нам не помешает. 

Теперь для каждого окна посчитайте 

$$Rolling Efficiency Ratio = \frac{Rolling Sum of Campaign Reach}{Rolling Sum of Ad Spend}$$

По сути, для каждого окна вы посчитаете сколько пользователе привлеклось за один доллар, потреченный на рекламу, в данном окне. Понятно, что значений будет столько, сколько окон. Нам интересно максимально значение такой эффективности для каждой платформы.

В качестве ответа выведите название платформы с наибольшей максимальной эффективность и наименьшей (два названия, не одно, не три, ровно два).

**Hint**: окна можно делать через [pd.DataFrame.rolling](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rolling.html).

In [11]:
# (◔/‿\◔)

df_sorted = df.sort_values(by=['Platform', 'Posts Per Week'], ascending=[True, False])

df_sorted['Rolling Reach'] = df_sorted.groupby('Platform')['Campaign Reach'].rolling(window=3, min_periods=1).sum().reset_index(level=0, drop=True)
df_sorted['Rolling Ad Spend'] = df_sorted.groupby('Platform')['Ad Spend (USD)'].rolling(window=3, min_periods=1).sum().reset_index(level=0, drop=True)

df_sorted['Rolling Efficiency Ratio'] = df_sorted['Rolling Reach'] / df_sorted['Rolling Ad Spend']

max_efficiency = df_sorted.groupby('Platform')['Rolling Efficiency Ratio'].max()
min_efficiency = df_sorted.groupby('Platform')['Rolling Efficiency Ratio'].min()

max_platform = max_efficiency.idxmax()
min_platform = min_efficiency.idxmin()

print(f"Платформа с наибольшей эффективностью: {max_platform}")
print(f"Платформа с наименьшей эффективностью: {min_platform}")

Платформа с наибольшей эффективностью: Facebook
Платформа с наименьшей эффективностью: Twitter


#### Задание 7

Это еще не все прекрасные функции pandas, которые мы хотим вам показать. Теперь вы посчитаете, сколько аккаунтов на каждой платформе одновременно лучшие по `Engagement Rate` и `Conversion Rate`.

Сделайте два отдельных суб-сета. В одном оставьте для каждой платфмормы один топовый аккаунт по `Engagement Rate`, в другом - по `Conversion Rate`. Соедините эти два подмножества по столбцу `Platform` так, что в одно строке есть описание сразу двух аккаунтов-лидеров. Теперь посмотрите равны ли имена аккаунтов в одной строке. Выведите количество строк, в которых названия аккаунтов совпадают.

In [12]:
# ( ͡° ͜ʖ ͡°)

top_engagement = df.loc[df.groupby("Platform")["Engagement Rate"].idxmax()]
top_conversion = df.loc[df.groupby("Platform")["Conversion Rate"].idxmax()]

merged_df = pd.merge(top_engagement, 
                     top_conversion, 
                     on="Platform", 
                     suffixes=("_eng", "_conv"))

matching_count = (merged_df["Username_eng"] == merged_df["Username_conv"]).sum()


print("Количество строк, в которых названия аккаунтов совпадают:", matching_count)

Количество строк, в которых названия аккаунтов совпадают: 0


#### Задание 8

Давайте теперь что-то попроще сделаем. Например, посчитаем отношение суммарного количества подписчиков на аккаунтах с высокой конверсией к такой же сумме в аккаунтах с низкой конверсией (очевидно, для каждой платформы). По сути, мы просто хотим получить число, которое характеризует, на сколько сильно аккаунты с высокой конверсией "доминируют" над аккаунтами с низкой конверсией в плане количества подписчиков.

Высокой конверсией будем считать конверсию больше средней. Остальное - низкая. Посчитайте суммы подписчиков для каждой платформы, поделите одно на другое и выведите разницу между самым большим значением и самым маленьким, а также платформы, которые соотвутствуют этим значениям.

Используйте магическую команду `%%time`, чтобы замерить, сколько времени ушло на исполнение вашего pandas-скрипта.

In [14]:
%%time
# (◡‿◡✿)

mean_conversion = df["Conversion Rate"].mean()

df["Conversion Level"] = df["Conversion Rate"].apply(lambda x: "high" if x > mean_conversion else "low")

followers_summary = df.groupby(["Platform", "Conversion Level"])["Follower Count"].sum().unstack()

followers_summary["Ratio"] = followers_summary["high"] / followers_summary["low"]

max_ratio = followers_summary["Ratio"].max()
min_ratio = followers_summary["Ratio"].min()

max_platform = followers_summary["Ratio"].idxmax()
min_platform = followers_summary["Ratio"].idxmin()

difference = max_ratio - min_ratio

print(f"Максимальное значение Ratio: {max_ratio:.2f} (Платформа: {max_platform})")
print(f"Минимальное значение Ratio: {min_ratio:.2f} (Платформа: {min_platform})")
print(f"Разница между максимумом и минимумом: {difference:.2f}")

Максимальное значение Ratio: 1.04 (Платформа: Twitter)
Минимальное значение Ratio: 0.86 (Платформа: Instagram)
Разница между максимумом и минимумом: 0.18
CPU times: total: 31.2 ms
Wall time: 24.2 ms


#### Задание 9

А теперь решите задание 8 чисто питоном. Никаких функций и методов pandas. Только питоновские циклы. Замерьте время выполнения кода. Наконец, сравните время в задании 8 и 9. Напишите ниже, кто же победил: чистый python и pandas?

**Hint**: Чтобы итерироваться по датафрейму, можно из него сделать генератор через [pd.DataFrame.iterrows](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iterrows.html) или [pd.DataFrame.itertuples](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.itertuples.html#pandas.DataFrame.itertuples). К слову, это не все способы итерироваться по датафрейму.

In [15]:
# (✿◠‿◠)

import time

start_time = time.time()

conversion_rates = [row["Conversion Rate"] for _, row in df.iterrows()]
mean_conversion = sum(conversion_rates) / len(conversion_rates)

# словарь для хранения подписчиков по платформам
followers_summary = {}

for _, row in df.iterrows():
    platform = row["Platform"]
    followers = row["Follower Count"]
    conversion = row["Conversion Rate"]
    
    level = "high" if conversion > mean_conversion else "low"

    if platform not in followers_summary:
        followers_summary[platform] = {"high": 0, "low": 0}
    followers_summary[platform][level] += followers

ratios = {platform: data["high"] / data["low"] for platform, data in followers_summary.items()}

max_platform = max(ratios, key=ratios.get)
min_platform = min(ratios, key=ratios.get)

difference = ratios[max_platform] - ratios[min_platform]

end_time = time.time()

print(f"Максимальное значение Ratio: {ratios[max_platform]:.2f} (Платформа: {max_platform})")
print(f"Минимальное значение Ratio: {ratios[min_platform]:.2f} (Платформа: {min_platform})")
print(f"Разница между максимумом и минимумом: {difference:.2f}")

print(f"Время выполнения (чистый Python): {end_time - start_time:.4f} секунд")


Максимальное значение Ratio: 1.04 (Платформа: Twitter)
Минимальное значение Ratio: 0.86 (Платформа: Instagram)
Разница между максимумом и минимумом: 0.18
Время выполнения (чистый Python): 2.0188 секунд


**А победителем является**: Конечно же, победил pandas! У него время выполнения 24.2 ms, в то время, как код на чистом python занял 2.0188 секунд. Это связано с тем, что pandas использует векторизированные операции и оптимизированный C-код.
Чистый Python будет медленнее, так как использует циклы, которые в Python менее эффективны.

Код на pandas оказался в ~83 раза быстрее, чем чистый Python! 🚀

#### Задание 10

Крайне серьезное задание. Отнеситесь к нему соответствующе. В ячейке ниже напишите ваш любимый анекдот или мем (только без баянов, окей?). Можно плохие. Помните, это задание на полный балл. Проверяющий работу ассистент должен улыбнуться.

Если вставляете картинку, то убедитесь, что вы ее не подгружаете локально. А то будет неудобно - потерять балл на этом задании, когда надо было выложить картинку на облако и прокинуть ссылку. И нет, нельзя сюда просто ссылку вставить. Либо ищите, как вставить картинку, либо смешной анекдот. Есть всего два стула - выбирайте...

In [16]:
# ‿( ́ ̵ _-`)‿

# 　∧＿∧
# （｡･ω･｡)つ━☆・*。
# ⊂　　 ノ 　　　・゜+.
# 　しーＪ　　　°。+ *´¨)
# 　　　　　　　　　.• ´¸.•*´¨) ¸.•*¨)
# 　　　　　　　　　　(¸.•´ (¸.•’*
#          .        ) )       (( (    )   
#  (     (   (  (  (\(        ))\)\  (()  
#  )\    )\  )\ )\: )(|     (((_)_()()(_) 
# (_)    | |((_)(_)_()\     |_ _|   \   \ 
# | |    | | _ \ V / -_)     | || - | |) |
# |_|    |_|___/\_/\___|    |___|_|_|___/ 

Жил один рыжий человек, у которого не было глаз и ушей. У него не было и волос, так что рыжим его называли условно.
Говорить он не мог, так как у него не было рта. Носа тоже у него не было.
У него не было даже рук и ног. И живота у него не было, и спины у него не было, и хребта у него не было, и никаких внутренностей у него не было. Ничего не было!
Так что непонятно, о ком идёт речь.
Уж лучше мы о нём не будем больше говорить